test/client/api.js

```diff
@@ -427,4 +426,24 @@ ClientApi.heartbeat = function (origin) {
   return (new ClientApi(origin)).doRequest('GET', origin + '/__heartbeat__')
 }
 
+function getQueryString (options) {
+  var qs = ''
```

Comment:
It's not worth spending too many cycles, but I do think we should simplify the logic here a bit.  Maybe starting with a default `qs="?"` and ending with a `return qs.length === 1 ? '' : qs` would be enough to tidy up the code a bit without introducing all the edge-cases you list above?